lib/widgets/action_sheet.dart

```diff
@@ -183,41 +147,158 @@ class MessageActionSheetCancelButton extends StatelessWidget {
-        reactionType: ReactionType.unicodeEmoji,
-        emojiCode: '1f44d',
-        emojiName: '+1',
-      );
-    } catch (e) {
-      if (!pageContext.mounted) return;
+  void _onMorePressed() {
+    // Dismiss current action sheet before opening emoji picker sheet.
+    Navigator.of(pageContext).pop();
 
-      switch (e) {
-        case ZulipApiException():
-          errorMessage = e.message;
-          // TODO(#741) specific messages for common errors, like network errors
-          //   (support with reusable code)
-        default:
-      }
+    showEmojiPickerSheet(pageContext: pageContext, message: message);
+  }
 
-      showErrorDialog(context: pageContext,
-        title: 'Adding reaction failed', message: errorMessage);
+  @override
+  Widget build(BuildContext context) {
+    assert(EmojiStore.popularEmojiCandidates.every(
+      (emoji) => emoji.emojiType == ReactionType.unicodeEmoji));
+
+    final zulipLocalizations = ZulipLocalizations.of(context);
+    final store = PerAccountStoreWidget.of(pageContext);
+    final designVariables = DesignVariables.of(context);
+
+    bool hasSelfVote(EmojiCandidate emoji) {
+      return message.reactions?.aggregated.any((reactionWithVotes) {
+        return reactionWithVotes.reactionType == ReactionType.unicodeEmoji
+          && reactionWithVotes.emojiCode == emoji.emojiCode
+          && reactionWithVotes.userIds.contains(store.selfUserId);
+      }) ?? false;
     }
+
+    return Container(
+      decoration: BoxDecoration(color: designVariables.contextMenuCancelBg),
+      child: Row(children: [
+        Flexible(child: Row(
+          spacing: 1,
+          children: List.unmodifiable(EmojiStore.popularEmojiCandidates.mapIndexed((index, emoji) {
+            final isSelfVoted = hasSelfVote(emoji);
+            return Flexible(child: InkWell(
+              onTap: () => _onReactionPressed(emoji: emoji, isSelfVoted: isSelfVoted),
+              splashFactory: NoSplash.splashFactory,
+              borderRadius: index == 0
+                ? const BorderRadius.only(topLeft: Radius.circular(7))
+                : null,
+              overlayColor: WidgetStateColor.resolveWith((states) =>
+                states.any((e) => e == WidgetState.pressed)
+                  ? designVariables.contextMenuItemBg.withFadedAlpha(0.20)
+                  : Colors.transparent),
+              child: Container(
+                width: double.infinity,
+                padding: const EdgeInsets.symmetric(vertical: 12, horizontal: 5),
+                alignment: Alignment.center,
+                color: isSelfVoted
+                  ? designVariables.contextMenuItemBg.withFadedAlpha(0.20)
+                  : null,
+                child: UnicodeEmojiWidget(
+                  emojiDisplay: emoji.emojiDisplay as UnicodeEmojiDisplay,
+                  notoColorEmojiTextSize: 20.1,
+                  size: 24))));
+          })))),
+        InkWell(
+          onTap: _onMorePressed,
```

Comment:
nit: "onFoo" for callbacks, "handleFoo" for the method that is called:
```suggestion
          onTap: _handleTapMore,
```

https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#naming-rules-for-typedefs-and-function-variables

Basically it's useful to have some convention for naming the two sides differently, because lots of places have one of each of these kinds of things in the same scope, both being about the same thing, and so without such a convention you end up with name collisions that have to be resolved by hacks like picking a less-good name for one of the two.